### DISCUSSION TITLE
[提问]关于引擎帧缓冲区的问题

### DISCUSSION THREAD

**Discussion by 523810185** [BoomingTech/Piccolo#273] [Q&A] [answered]
想问问大佬们，关于Piccolo引擎wiki中说的帧缓冲区相关的问题，文档中说，设置缓冲区是为了保证所有依赖B组件获取到B的数据是一样的；但是看了代码中，transformCom的tick中交换了缓冲区，这样不是还是会导致，比B先更新的A获取的是缓冲区0，比B后更新的会获取缓冲区1的数据，两边数据不一样嘛？感觉交换缓冲区是不是应该放在帧尾，所有tick做完以后在做才比较符合需求

![1655453256](https://user-images.githubusercontent.com/43307388/174256418-8baea2db-7f59-47c3-af22-6c0f086418f4.png)
![1655453305](https://user-images.githubusercontent.com/43307388/174256422-a9a33881-948b-46ce-aee3-f7248bb015ad.png)

**Comment by Ol6rin** [comment, marked as answer]
同学你好，这里wiki文档之前描述的确实不够准确，已经修改为更明确的内容。

> GO内的组件依赖，如A、B、C、D 四个组件依次 tick，B、D两个组件都会读取A组件中的状态，C组件 tick 时会修改A组件的状态（如Motor组件更新了GO的变换）。不做任何处理的情况下会出现B组件使用的A组件的状态是这一帧更新之前的，D组件使用的A组件状态是这一帧更新之后的，会引发逻辑bug。

> Piccolo引擎中解决这个问题的方案是帧缓冲区，给A组件的状态提供两个缓冲区，每一帧计算时一个缓冲区作为当帧输入缓冲区，另一个缓冲区作为当帧输出缓冲区，以保证所有依赖D组件的其他组件在这一帧获取的输入数据都是一样的。A组件tick时交换两个缓冲区，并将上一帧的输入缓冲区复制进下一帧的输出缓冲区，以保证A组件在某一帧停止更新时，不会在下一帧跳回至上一帧的状态。另一方面将被依赖的组件安排在了前面，保证不会在一帧当中交换缓冲区。

这里隐含了一个条件是，被依赖的组件组织的时候应该放在前面，否则确实应该在帧尾做交换。
